### DISCUSSION TITLE
How to access Resource in an async task?

### DISCUSSION THREAD

**Discussion by feelingsonice** [bevyengine/bevy#11251] [Q&A]
I have a procedural generation task that may take a long time to complete so I've create a plugin to listen for an `Event` trigger. The task requires access to certain `Resource` stored as either an `Asset` resource or plain `Resource`. This means I only have borrowed access to it AFAIK since Bevy stores and owns it. 

I'd really rather not clone this `Resource`. Ideally, this event listener is able retrieve some sort of reference to it, pass it to the async task (I'm using `AsyncComputeTaskPool`), and that's it. However, I can't send the `Res` via the `Event` as there's lifetime issues, and I'm also not seeing a easy way to pass it **into** the async task.

Anyone have suggestions?

**Comment by TimJentzsch** [comment]
The only safe way to do this that I can think of is to wrap the contents of your `Resource` in an `Arc<Mutex<T>>`.
The `Arc` can be cheaply cloned (without cloning `T`) and the `Mutex` ensures that the async task and the Bevy systems cannot access `T` at the same time.
Otherwise you could easily run into memory issues.

Just keep in mind that a `Mutex` of course adds overhead which might slow down your app, depending on your use cases.

**Reply by feelingsonice** [reply]
I'm planning to just put it behind arc and skip the Mutex since I don't need to mutate this at all. You think that'd work?

**Reply by nicopap** [reply]
The downside is that you won't be able to mutate the resource as long as an async task is running, though you can still read it. You can use [`Arc::get_mut`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.get_mut) if the `Arc` inside the resource is truly the last reference existing.
